doc/source/reference/sparse.migration_to_sparray.rst

```diff
@@ -0,0 +1,268 @@
+.. _migration_to_sparray:
+
+Migration from spmatrix to sparray
+==================================
+
+This document provides guidance for converting code from sparse matrices
+to sparse arrays in SciPy Sparse.
+
+The change from sparse matrices to sparse arrays mirrors conversion from
+``np.matrix`` to ``np.array``. Essentially we must move from an all-2D
+matrix multiplication centric ``matrix`` object to a 1D or 2D “array”
+object that supports operand matrix multiplication and elementwise
+computation.
+
+Notation: For this guide we denote the sparse array classes generally as
+``sparray`` and the sparse matrix classes ``spmatrix``. Dense numpy
+arrays are denoted ``np.array`` and dense matrix classes are
+``np.matrix``. Supported sparse formats are denoted BSR, COO, CSC, CSR,
+DIA, DOK, LIL and all formats are supported by both sparray and
+spmatrix. The term ``sparse`` refers to either ``sparray`` or
+``spmatrix``, while ``dense`` refers to either ``np.array`` or
+``np.matrix``.
+
+Overview and big picture:
+-------------------------
+
+-  the constructor names ``*_matrix``, e.g. ``csr_matrix``, are changed
+   to ``*_array``.
+-  spmatrix ``M`` is always 2D (rows x columns) even
+   e.g. ``M.min(axis=0)``. sparray ``A`` can be 1D or 2D. Numpy scalars
+   are used as 0D values where needed.
+-  Operators change: ``*, @, *=, @=, **``
+
+   -  scalar multiplication uses ``*`` and is not implemented for ``@``.
+   -  sparrays use ``*`` for elementwise multiplication and ``@`` for
+      matrix multiplication while spmatrices can use both operators
+      ``*`` and @ for matrix multiplication. Either can use
+      ``A.multiply(B)`` for elementwise multiplication.
+   -  scalar exponents ``A**2`` use elementwise power for sparray and
+      matrix power for spmatrix. To get matrix power for arrays use
+      ``sp.sparse.linalg.matrix_power(A, n)``
+
+-  Checking the sparse type and format:
+
+   -  ``issparse(A)`` returns ``True`` for any sparse array/matrix.
+   -  ``isspmatrix(M)`` returns ``True`` for any sparse matrix.
+   -  ``issparse(A) and A.format == 'csr'`` looks for a CSR sparse
+      array/matrix
+
+Recommended steps for migration:
+--------------------------------
+
+-  First Pass (leaving spmatrix in the code)
+
+   -  In your spmatrix code, change ``*`` to ``@`` for matrix
+      multiplication. Note that scalar multiplication with sparse should
+      use ``*``.
+   -  scalar powers, e.g. ``M**3``, should be converted to
+      ``sp.sparse.linalg.matrix_power(M, 3)`` or ``M @ M @ M`` if the
+      power is known.
```

Comment:
Presumably the first call is faster than the second.

Reply:
I removed the slower multiple calls to the `@` operator to not confuse things.